selinux: Do parallel relabel on package install

b-ranto opened the pull request:
We can take advantage of ceph-disk fix subcommand when doing a package
install. We will keep using the differential fixfiles command otherwise.

We also need to add relabel for /usr/bin/ daemons so that we could use
this.

Signed-off-by: Boris Ranto <branto@redhat.com>

This should help when we are upgrading from a non-SELinux enabled ceph version to a SELinux enabled ceph version since we will be doing relabel in parallel per OSD.

b-ranto commented:
@ktdreyer @smithfarm could you review? The idea here is quite simple -- if we are doing full relabel we might as well use the ceph-disk fix for that to relabel in parallel per OSD.

ktdreyer commented:
So `fixfiles` runs in parallel but `ceph-disk --selinux` does not?

b-ranto commented:
@ktdreyer the other way around, `ceph-disk` runs `restorecon` in parallel per osd but fixfiles does not, it is a post script so if $1 = 1 it is a first installation of the package, not an update (correct me if I am wrong, here)

b-ranto commented:
Could you review/merge? This is especially helpful when people are upgrading from a non-SELinux enabled ceph version to an SELinux enabled one and they are running with SELinux enabled (be it in Permissive or Enforcing mode).

smithfarm commented:
I have looked at this, and it seems OK to me, but don't know enough about the subject to approve the PR.

smithfarm commented:
@b-ranto Run it through a `ceph-disk` suite?

dachary commented:
@smithfarm there are no tests covering that part of the code

ktdreyer commented:
Would you please file a http://tracker.ceph.com/ ticket for this to ensure it gets into jewel?

b-ranto commented:
@smithfarm The only test we have for that part of the code is run by the build itself.

@ktdreyer done, I have also updated the commit messages to include the Fixes line

smithfarm commented:
Reviewed-by: Loic Dachary <ldachary@redhat.com>
Reviewed-by: Ken Dreyer <kdreyer@redhat.com>

smithfarm commented:
@b-ranto ceph-disk unit test failure?

b-ranto commented:
@smithfarm Yeah, I had to update the unit test because now, the code is testing for presence of the files to restorecon/chown and if ceph is not installed on the machine doing the build, it won't do any restorecon/chown so the assert looking for these fails.

b-ranto commented:
QE raised a very good point that ceph-disk binary is not available on monitors (i.e. without ceph-osd package installed). We are already shipping the ceph-disk library in ceph-base so moving ceph-disk to ceph-base seems like a reasonable next step. I have rebased the branch and addressed that in the latest commit. Re-review the commit please.

ktdreyer commented:
Using ceph-disk to relabel monitors feels really weird to me

alfredodeza commented on ceph.spec.in:
```diff
@@ -1719,8 +1730,14 @@ if test $STATUS -eq 0; then
+# Relabel the files
+# Use ceph-disk fix for first package install and fixfiles otherwise
+if [ "$1" = "1" ]; then
+    /usr/sbin/ceph-disk fix --selinux
```
@b-ranto it doesn't look like ceph-disk does too much to make this happen, would it be possible to just run the same shell command here instead? Or what was the difficulty you saw with that approach?

b-ranto replied:
The ceph-disk command does the relabel in parallel per osd which can speed up the relabel a lot as the OSDs usually reside on different disks.

b-ranto commented:
@ktdreyer it relabels all of ceph data including mon/mds/rgw/... data

b-ranto commented:
jenkins retest this please

smithfarm commented:
Will also need https://github.com/ceph/ceph/pull/16208 once that PR gets merged into master.
